packages/mui-base/src/useNumberInput/useNumberInput.types.ts

```diff
@@ -0,0 +1,220 @@
+import * as React from 'react';
+import { FormControlState } from '../FormControl';
+
+export type UseNumberInputChangeHandler = (
+  e: React.KeyboardEvent<HTMLInputElement>,
+  value: number | null,
+) => void;
+
+export interface UseNumberInputParameters {
+  /**
+   * The minimum value.
+   */
+  min?: number;
+  /**
+   * The maximum value.
+   */
+  max?: number;
+  /**
+   * The amount that the value changes on each increment or decrement.
+   */
+  step?: number;
+  /**
+   * Multiplier applied to `step` if the shift key is held while incrementing
+   * or decrementing the value. Defaults to `10`.
+   */
+  shiftMultiplier?: number;
+  /**
+   * The default value. Use when the component is not controlled.
+   */
+  defaultValue?: unknown;
+  /**
+   * If `true`, the component is disabled.
+   * The prop defaults to the value (`false`) inherited from the parent FormControl component.
+   */
+  disabled?: boolean;
+  /**
+   * If `true`, the `input` will indicate an error by setting the `aria-invalid` attribute.
+   * The prop defaults to the value (`false`) inherited from the parent FormControl component.
+   */
+  error?: boolean;
+  onBlur?: React.FocusEventHandler<HTMLInputElement>;
+  onClick?: React.MouseEventHandler;
+  /**
+   * Callback fired when the <input> value changes, before clamping is applied. Note that
+   * `event.target.value` may contain values that fall outside of `min` and `max` or
+   * are otherwise "invalid".
+   */
+  onChange?: React.ChangeEventHandler<HTMLInputElement>;
+  onFocus?: React.FocusEventHandler;
+  /**
+   * Callback fired after the value is clamped and changes.
+   * Called with `undefined` when the value is unset.
+   */
+  onValueChange?: (
```

Comment:
Extra relevant data points: #14674

Reply:
> it seems that both events are fired at the same time.

There is a difference! When using the textbox, `onValueChange` fires on blur as it has to take clamping into account (when using the buttons they do fire at the same time) – let me rewrite the description to be more clear for the API docs

> However, https://mui.com/material-ui/react-autocomplete/#controlled-states sounds like a more relevant prior art

This was actually my first idea, but because `onValueChange` fires on blur and not with every keystroke, I thought that calling it `onChange` may cause some initial confusion, as it behaves quite differently from the "typical" onChange that you would expect on a textbox

> this seems to be the value most developers will use, assuming that value control this.

This is correct, and may be a good reason for calling it `onChange` instead, so far I've tried to keep it as close to the native one as possible. This way would also avoid introducing another custom change handler (onValueChange) for users to remember, what are your thoughts @mnajdova @michaldudak

Reply:
> onValueChange -> onChange this seems to be the value most developers will use, assuming that value control this.
> onChange -> onInputChange this is the value of the input, we could later introduce a inputValue prop if developers find use cases for it.

👍 agree, this makes most sense to me too.